master/internal/proxy/proxy.go

```diff
@@ -52,54 +29,68 @@ type ProxyHTTPAuth func(echo.Context) (done bool, err error)
 // Proxy is an actor that proxies requests to registered services.
 type Proxy struct {
 	lock     sync.RWMutex
+	HTTPAuth ProxyHTTPAuth
 	services map[string]*Service
+	syslog   *logrus.Entry
+}
 
-	HTTPAuth ProxyHTTPAuth
+// DefaultProxy is the global proxy singleton.
+var DefaultProxy *Proxy
+
+// InitProxy initializes the global proxy.
+func InitProxy(httpAuth ProxyHTTPAuth) {
```

Comment:
haven't figured out how to make this `(p *Proxy) InitProxy` without getting warnings for possible nil pointer dereferences

Reply:
I think this makes sense as is. ive been putting the singletons and init func in a `singleton.go` file and then having, e.g., proxy.go just being the proxy struct and a New func to construct it

Reply:
actually liked Max's implementation here https://go.dev/play/p/GW6XR221CUY -- and I think it satisfies what you're looking for too!

Reply:
eh, personally i dont like structs with `(t *T) Init(...) ...`, i think a `NewT(...) *T` is a more clear API for 'how to construct a correct `T`.

Reply:
but i dont really care a whole lot, since in this case it is only used in one place

Reply:
> `NewT(...) *T` is a more clear API for 'how to construct a correct `T`.

Agreed. This function's existence signals that users shouldn't try to create their own `T` and then there's not an intermediate state where you have a `T` but it's not usable.